### ISSUE TITLE
Language picker inaccessibility

### ISSUE THREAD

**Issue by bbsmooth** [freeCodeCamp/freeCodeCamp#45106]
## Affected page

Language picker in main menu for curriculum.

## Discussion

The language picker in the main menu in the curriculum is currently inaccessible to keyboard only users because the change is triggered as soon as you press the down arrow key (which triggers a change event causing the language to be changed immediately). There really is no good way to use a `<select>` here if you want the change to be triggered without clicking a separate button to initiate the change (this just isn't how selects were intended to operate). You basically have to implement an alternative solution that is accessible to keyboard users.

I have spent the past few days creating an alternative, accessible solution, and am just about ready to submit the PR, but before I do that I can't help thinking that this might be over-complicating the issue and there is a better way. I'd like to suggest that we move the language picker to the user's settings page. If we did that then we could have a button that initiates the change and then we could use a plain old `<select>` which would reduce the complexity of implementing this functionality considerably. We could still include the language picker in the main menu, but it would just be a link that takes the user to their settings page.

**Comment by bbsmooth** [comment]
OK, I just discovered that you can actually use the keyboard to force the select list to show all of the options and then arrow to the option you want and hit 'Enter' to select it. (On my computer usng FF it is `Alt+DownArrow). So there is in fact a way to make the language select work as it is now and thus it is technically not inaccessible to keyboard only users.

I think I would still argue that this workaround may not be known to all keyboard users, so having this option on the settings page instead would allow us to make it a little easier for keyboard users to use. But if we are happy with keeping it on the main menu for now then we are good to go for accessibility and all the fun I had implementing a custom language picker was for naught.

**Comment by bbsmooth** [comment]
Sorry for all the updates, I'm still learning some of this accessibility stuff.

After further investigation, I think the current language picker implementation is actually not technically accessible even though the user can use the "Alt+Arrow" combination to expand the select and choose a language. According to [WCAG 3.2.2 On Input](https://www.w3.org/WAI/WCAG21/Understanding/on-input.html):

"Changing the setting of any user interface component does not automatically cause a change of context unless the user has been advised of the behavior before using the component."

The "change of context" in this case is changing to a new language. Since we aren't warning the user that the mere act of selecting a language in the select will automatically cause the page to refresh in that language then we are in violation of 3.2.2. Adding a button that says "update language" would remediate this issue because the button is telling the user what will happen as a result of clicking it. The other option here would be to add visible text that tells the user that selecting a language will change the language of the page immediately, which I don't think would fit nicely in the menu.

So I guess I am reinstating by original suggestion that we move the language picker to the settings page and have a link to it in the main menu.

**Comment by huyenltnguyen** [comment]
@bbsmooth Thanks for opening the issue, and no need to apologize. I love reading these analyses and I actually have learned a ton from your opened issues and PRs (case in point: this a11y requirement). 

Accessibility aspect aside, I think language is not something the user would need to change often, so moving the picker to the Setting page would allow us to preserve some space in the dropdown menu. The Account Settings section seems to be the most appropriate place to add the picker, but @ahmadabdolsaheb is the best person for UI/UX recommendation.

Additionally, we should add the `lang` attribute to the options to help screen readers announce them properly (I just tested with VoiceOver and it is struggling with Chinese, Ukrainian, and Japanese).

**Comment by ShaunSHamilton** [comment]
Thanks, for taking the time to debug/research this.

**To Stress**: Some of the main _features_ of the freeCodeCamp community are:
a) The ability to go through the curriculum without an account
b) The wonderful effort that has gone into making the content accessibility in other languages

So, we have the language setting there, because we want non-signed in Campers (no access to a `/settings` page) to be able to access the language.

> if you want the change to be triggered without clicking a separate button to initiate the change

Personally, I dislike this current behaviour.

**My suggestion**

The dropdown could appear as though navigating with a mouse, and only upon selection (`enter`), does the languages get selected, and the page change location.

**Comment by bbsmooth** [comment]
Sorry, I was under the impression that the language option only showed up in the main menu if you were signed in. I see now that it is always there (my bad). If I had realized this earlier I wouldn't have suggested we move it to the settings page. Seeing as this needs to stay available even when not logged in then yes, it makes sense to keep it in the main menu. 

> The dropdown could appear as though navigating with a mouse, and only upon selection (`enter`), does the languages get selected, and the page change location.

This is exactly what my alternative version does. Basically, you have to roll your own custom selection menu so that the change does not trigger until the user hits the Enter button.

The other option would be to pop up a modal for the language picker, which would allow us to use a simple `<select>` with a `<button>` to trigger the change. I'm starting to think this might be the better approach because it would make it clear that the language will change immediately when clicking the button (satisfying WCAG 3.2.2) and we could also rely on default browser positioning when the select is expanded rather than having to position it ourselves when there isn't enough height in the view port.

**Comment by ShaunSHamilton** [comment]
@bbsmooth That all sounds fine to me. Except, I know people are buy-n-large anti-modal.

Is it bad practice to just Jerry-rig a group of `input type="radio"` for this?

**Comment by bbsmooth** [comment]
By default, radio buttons aren't intended to trigger a context change by pressing the Enter when they are focused, so I think that would be a big anti-pattern. On the [WAI-ARIA best practices example for radio buttons](https://www.w3.org/TR/wai-aria-practices/examples/radio/radio-activedescendant.html#kbd_label) they don't even define functionality for the Enter key. My alternative uses a [custom navigation submenu](https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-links.html#ex_label) which does specify that the Enter/Space keys activate the menu item. I'm still not convinced this 100% meets WCAG 3.2.2, but it is definitely more accessible than what we currently have.

Modals get a bad rap because they often aren't implemented properly (i.e. accessibly). I can assure you that mine would be as accessible as possible :-) Does FCC currently use any modals? I don't believe I've run into any yet but maybe I'm not looking in the right place?

**Comment by ShaunSHamilton** [comment]
> Does FCC currently use any modals? I don't believe I've run into any yet but maybe I'm not looking in the right place?

Only every time you complete a challenge/project.

Are we talking about different things, when we talk about modals? I mean the "fake page" that pops up taking focus - I dislike them on a usability standpoint, but also mostly because most that I have encountered are ads.

**Comment by ahmadabdolsaheb** [comment]
Thanks @bbsmooth for looking into improving the accessibility of the language picker.

We have placed the language picker in the main navigation to increase awareness for our i18n clients  and our  translation efforts.

**Comment by bbsmooth** [comment]
> If we change the "English" in the menu to a button with area "Change Language", it should do the trick?

You are reading my mind. That's exactly what I did (before I even read this).
